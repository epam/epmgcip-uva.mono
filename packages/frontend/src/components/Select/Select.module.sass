@use 'src/style/colors'
@import 'src/style/typography'

.label
    @extend .H6
    display: flex
    flex-direction: column

.selectRequired
    color: colors.$orange

.select
    @extend .T1
    width: calc(100vw - 48px)
    max-width: calc(768px - 48px)
    height: 48px
    border: 1px solid colors.$gray
    border-radius: 8px
    background-color: colors.$white
    padding: 12px
    box-sizing: border-box
    margin-top: 8px
    margin-bottom: 18px
    appearance: none
    background-image: url("./assets/arrow.svg")
    background-repeat: no-repeat
    background-position: right 21px top 50%

    &:active
        opacity: 0.75

.select select
    display: none

.placeholder
    @extend .T1
    color: colors.$gray