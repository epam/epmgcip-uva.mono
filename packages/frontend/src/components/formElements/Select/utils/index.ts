import { IOption } from "src/types";

export const setMultipleValue = <T,>(previousValue: T, newValue: T): T => {
    if (!Array.isArray(previousValue)) return newValue;
  
    if (previousValue.includes(newValue)) {
      return previousValue.filter((element) => element !== newValue) as T;
    }
  
    return [...(previousValue as T[]), newValue].sort() as T;
  };
  
  export const isValueSelected = <T,>(allValues: T, value: T): boolean =>
    Array.isArray(allValues) ? allValues.includes(value) : allValues === value;
  
  export const showCurrentValues = <T,>(
    options: IOption[],
    currentValue: T,
    multiple: boolean,
    placeholder: string
  ) => {
    if (multiple && Array.isArray(currentValue)) {
      const values = [] as T[];
  
      options.forEach((option) => {
        if (currentValue.includes(option.value)) {
          return values.push(option.name as T);
        }
      });
  
      return values.length === 0 ? placeholder : values.join(', ');
    }
  
    return (
      options.find((option) => option.value === currentValue)?.name || placeholder
    );
  };