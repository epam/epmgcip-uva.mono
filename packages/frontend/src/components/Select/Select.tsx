import { getClassesList } from 'src/utils';
import css from './Select.module.sass';

interface IOption {
  name: string;
  value: string;
}

interface SelectProps<T> {
  value: string;
  setChange: React.Dispatch<React.SetStateAction<T>>;
  options: IOption[];
  labelText: string;
  required?: boolean;
  placeholder?: string;
  labelClassName?: string;
  selectClassName?: string;
  isValidationError?: boolean;
  errorMessage?: string;
}

export const Select = <T,>({
  value,
  setChange,
  options,
  labelText,
  required = false,
  placeholder,
  labelClassName,
  selectClassName,
  isValidationError,
  errorMessage,
}: SelectProps<T>) => {
  const labelClasses = getClassesList(css.label, labelClassName);
  const selectClasses = getClassesList(
    css.select,
    selectClassName,
    isValidationError ? css.selectError : undefined
  );

  return (
    <label className={labelClasses}>
      <div>
        {labelText}
        {required && <span className={css.selectRequired}> *</span>}
      </div>
      <select
        value={value}
        onChange={(e) => setChange(e.target.value as T)}
        className={selectClasses}
      >
        {placeholder && (
          <option disabled={true} value=''>
            {placeholder}
          </option>
        )}
        {options.map(({ name, value }) => (
          <option key={value} value={value}>
            {name}
          </option>
        ))}
      </select>
      {isValidationError && (
        <p className={css.validationError}>
          {isValidationError && errorMessage}
        </p>
      )}
    </label>
  );
};
