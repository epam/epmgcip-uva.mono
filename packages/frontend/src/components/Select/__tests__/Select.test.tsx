import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { Select } from '../Select';

const valueMock = 'Test Value';
const setChangeMock = vi.fn();
const optionMock = [
  {
    name: 'Name 1',
    value: 'value_1',
  },
  {
    name: 'Name 2',
    value: 'value_2',
  },
];
const labelTextMock = 'Test Label';

describe('Testing: Select', () => {
  it('should render', () => {
    const { container } = render(
      <Select
        value={valueMock}
        setChange={setChangeMock}
        options={optionMock}
        labelText={labelTextMock}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
