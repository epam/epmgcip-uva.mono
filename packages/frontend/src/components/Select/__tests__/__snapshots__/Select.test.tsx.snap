// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: Select > should render 1`] = `
<div>
  <label
    class="_labelClass_1e72f6"
  >
    <div>
      Test Label
    </div>
    <select
      class="_select_1e72f6"
    >
      <option
        value="value_1"
      >
        Name 1
      </option>
      <option
        value="value_2"
      >
        Name 2
      </option>
    </select>
  </label>
</div>
`;
